Realm/Realm/Native/SyncError.cs

```diff
@@ -0,0 +1,49 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System.Runtime.InteropServices;
+using Realms.Sync.Exceptions;
+using Realms.Sync.Native;
+
+namespace Realms.Native
+{
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct SyncError
+    {
+        public ErrorCode error_code;
+
+        public PrimitiveValue message;
+
+        public PrimitiveValue log_url;
+
+        [MarshalAs(UnmanagedType.U1)]
+        public bool is_client_reset;
+
+        public MarshaledVector<StringStringPair> user_info_pairs;
+
+        public MarshaledVector<CompensatingWriteInfo> compensating_writes;
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct CompensatingWriteInfo
+    {
+        public PrimitiveValue reason;
+        public PrimitiveValue object_name;
```

Comment:
I really don't like we use a whole `PrimitiveValue` when we just need a string 😛
I've been working on a new `String` struct for native marshaling for the websocket code we could use instead.

Reply:
Isn't a "whole" `PrimitiveValue` like 17 bytes instead of the 16 (on 64 bit platforms) you would otherwise use 😄 We also have a [`StringValue`](https://github.com/realm/realm-dotnet/blob/547ba96b29613fbe2e525bc0c974835efafa87da/Realm/Realm/Native/PrimitiveValue.cs#L285-L290) type which I could have used instead, but was too lazy to move some code around. As far as I can tell, it's the same as the `String` struct you're proposing in your PR.

Reply:
Oh, you're right - I missed `StringValue`.

I don't think using `PrimitiveValue` here is a good idea, because I suspect marshaling a union is more expensive as .NET doesn't have "unions" the way C understands them.

Reply:
Okay, it was bugging me anyway that we're not more explicit about the field types. I replaced usages with `StringValue`.